packages/docusaurus-theme-common/src/contexts/doc.tsx

```diff
@@ -0,0 +1,63 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, {useMemo, type ReactNode, useContext} from 'react';
+import {ReactContextError} from '../utils/reactUtils';
+import type {PropDocContent} from '@docusaurus/plugin-content-docs';
+
+export type DocContextValue = {MDXComponent: () => JSX.Element} & Pick<
+  PropDocContent,
+  'metadata' | 'frontMatter' | 'toc' | 'assets' | 'contentTitle'
+>;
+
+const Context = React.createContext<DocContextValue | null>(null);
+
+/**
+ * Note: we don't use PropDoc as context value on purpose
+ * Metadata is currently stored inside the MDX component,
+ * but we may want to change that in the future.
+ * This layer is a good opportunity to decouple storage from consuming API.
+ */
+function useContextValue(content: PropDocContent): DocContextValue {
```

Comment:
My opinion: storing metadata/frontmatter in the mdx file is an implementation detail. Consuming the data with `Component.metadata` doesn't feel very natural to me so I'd rather not stick to it 🤷‍♂️ 

In practice, we could create separate bundles/props for these metadata. And it may have advantages to do so in the future if for example we want to provide some kind of "data query layer" (I mean, for example we want to display some docs/blog metadata in the homepage, without having to load all the mdx content in that page)

I'm not sure how things will turn out, but I'd prefer to decouple the metadata from the MDX component when consuming it. I'd like to do something similar on the blog as well.

Reply:
I agree—the metadata should be a separate chunk. This means we can easily provide the same metadata as both props and context. Seems too complex in the short term, though.